File: src/328.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    '''
    感觉应该使用分治算法，加上双指针
    是一种穿针引线的感觉
    不分治，两个指针就搞定了，还比比来来那么多
    time defeat: 97.79%%
    space defeat: 76.57%
    '''
    def oddEvenList(self, head: ListNode) -> ListNode:
        #短路原则，不会报错
        if head == None or head.next == None or head.next.next == None:
            return head
        oddPointer = head   #奇数指针
        evenPointer = head.next #偶数指针
        #记录奇偶数链表的头和尾
        evenHead = head.next
        while evenPointer and evenPointer.next:
            #不拼接，直接改
            oddPointer.next = evenPointer.next
            oddPointer = oddPointer.next
            evenPointer.next = oddPointer.next
            evenPointer = evenPointer.next
        oddPointer.next = evenHead
        return head

File: src/328.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        //干翻字节！
        //题目要求空间O1，时间On
        //链表题，用c++写有感觉
        //第一个为奇数，第二个为偶数
        //先找到第一个和第二个节点
        //这个题比较简单，十分钟，就这还中等题？
        //一次提交，时间击败-88.79%，空间击败-49.59%
        if(head == nullptr || head->next == nullptr){
            return head;
        }
        ListNode * oddNode = head;
        ListNode * evenNode = head->next;
        //保存偶数节点头
        ListNode * evenNodeHead = evenNode;
        //开始串联，一个为空就结束
        while(oddNode && evenNode){
            oddNode->next = evenNode->next;
            oddNode = oddNode->next;
            if(oddNode){
                evenNode->next = oddNode->next;
                evenNode = evenNode->next;
            }
        }
        //连接奇数和偶数链表
        ListNode * temp = head;
        while(temp->next){
            temp = temp->next;
        }
        temp->next = evenNodeHead;
        return head;
    }
};
